numba/cuda/tests/cudapy/test_frexp_ldexp.py

```diff
@@ -0,0 +1,52 @@
+import numpy as np
+import math
+from numba import cuda
+from numba.cuda.testing import unittest, CUDATestCase
+
+
+def simple_frexp(aryx, aryexp, arg):
+    aryx[0], aryexp[0] = math.frexp(arg)
+
+
+def simple_ldexp(aryx, arg, exp):
+    aryx[0] = math.ldexp(arg, exp)
+
+
+class TestCudaFrexpLdexp(CUDATestCase):
+    def test_frexp_f4(self):
```

Comment:
You don't want to test against `inf` and `nan`, because `frexp` gives unspecified results in those cases https://en.cppreference.com/w/c/numeric/math/frexp

Reply:
In the C++ specification, it looks that the exponent is unspecified in those cases. CPython deals with these cases explicitly, always returning the input for the fractional part and 0 for the exponent: https://github.com/python/cpython/blob/master/Modules/mathmodule.c#L2150 - Numba aims to replicate CPython semantics, so we should do what CPython does. I'd be happy with either:

- Changing the lowering of `math.frexp` so that it always returns `(x, 0)` for infinite / NaN `x`, and testing with `inf` and `nan`, or
- Adding a note to the documentation stating that the exponent is unspecified for infinite / NaN values, which is a divergence from CPython.

I think my preference would be towards the former - what are your thoughts on the above?

Reply:
I think I should take a look at what `__nv_frexp` returns first.  Because I suspect that it actually returns a specific value.  The exp may be unspecified only because `frexp` is also a part of math API that works on CPU.

Reply:
OK - sounds good!

Reply:
So, luckily, `__nv_frexp` and Python's `math.frexp` behave the same.  They both return `(inf, 0)` and `(nan, 0)` for `inf` and `nan`, respectively.  This is different from `np.frexp` which sets exponents to `-1`, but we are not lowering that in this patch.